Improve `BindingSource` values when custom model binders are in use

dougbu opened the pull request:
- #4939
- make it easy to select `BindingSource` in `[BindProperty]` and `[ModelBinder]`
- change default when `BinderType` is set to `BindingSource.ModelBinding`

smaller issues:
- catch invalid `BinderType` values up front
- complete `BindingSource.ModelBinding` implementation: `IValueProvider` filtering was faulty

nits:
- accept VS suggestions e.g. remove unused variables
- add `[BindProperty]` doc comments
- add `<remarks>` to `BinderType` properties that recommend setting `BindingSource` in some cases
- "model binder" -> `<see cref="IModelBinder" /> implementation` in some doc comments

dougbu committed:
Improve `BindingSource` values when custom model binders are in use
- #4939
- make it easy to select `BindingSource` in `[BindProperty]` and `[ModelBinder]`
- change default when `BinderType` is set to `BindingSource.ModelBinding`

smaller issues:
- catch invalid `BinderType` values up front
- complete `BindingSource.ModelBinding` implementation: `IValueProvider` filtering was faulty

nits:
- accept VS suggestions e.g. remove unused variables
- add `[BindProperty]` doc comments
- add `<remarks>` to `BinderType` properties that recommend setting `BindingSource` in some cases
- "model binder" -> `<see cref="IModelBinder" /> implementation` in some doc comments

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/BindingInfo.cs:
```diff
@@ -48,9 +52,29 @@ public BindingInfo(BindingInfo other)
+            get => _binderType;
+            set
+            {
+                if (!(value == null || typeof(IModelBinder).GetTypeInfo().IsAssignableFrom(value.GetTypeInfo())))
```
`if (value != null && !typeof(IModelBinder).IsAssignableFrom(value))`

dougbu replied:
Think I copied the `GetTypeInfo()` use from `BinderTypeModelBinder`. Will fix that up too.

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/BindingSource.cs:
```diff
@@ -196,7 +196,22 @@ public virtual bool CanAcceptDataFrom(BindingSource bindingSource)
+                return bindingSource == Form || bindingSource == Path || bindingSource == Query;
+            }
+
+            if (bindingSource == ModelBinding)
```
This condition seems really strange. The previous if statement makes intuitive sense, but in combination with it, it suggests that `Path`, `Query` and `Form` are sort of aliases for `ModelBinding`. Could you add a comment that says why this condition is required?

dougbu replied:
Good catch. I'll remove https://github.com/aspnet/AspNetCore/blob/e0de8a31375e173c03ddb5cd7f21bd5bcc252412/src/Mvc/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/BindingSource.cs#L209-L212

dougbu committed:
Address PR comments

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/BindingSource.cs:
```diff
@@ -196,7 +196,17 @@ public virtual bool CanAcceptDataFrom(BindingSource bindingSource)
+                return bindingSource == Form || bindingSource == Path || bindingSource == Query;
+            }
+
+            return false;
```
Was this a bug?

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/BindingSource.cs:
```diff
@@ -220,9 +230,9 @@ public override int GetHashCode()
         public static bool operator ==(BindingSource s1, BindingSource s2)
         {
-            if (object.ReferenceEquals(s1, null))
+            if (s1 is null)
```
ewww

pranavkm replied:
💯 points

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.ApiExplorer/DefaultApiDescriptionProvider.cs:
```diff
@@ -278,8 +278,7 @@ private void ProcessRouteParameters(ApiParameterContext context)
             {
                 if (parameter.Source == BindingSource.Path ||
-                    parameter.Source == BindingSource.ModelBinding ||
-                    parameter.Source == BindingSource.Custom)
```
Cool 👍

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs:
```diff
@@ -2,31 +2,102 @@
+                throw new InvalidEnumArgumentException(nameof(bindingSource), sourcesIndex, typeof(BindingSourceKey));
+            }
+
+            _bindingSource = BindingSources[sourcesIndex];
```
Everything in this constructor should probably be a static method on `BindingSource`. It's a little bit random to find it implemented here.

rynowak commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/BindPropertyAttribute.cs:
```diff
@@ -2,31 +2,102 @@
+        /// The <see cref="BindingSourceKey"/> that indicates the value of the
+        /// <see cref="BindingSource"/> property.
+        /// </param>
+        public BindPropertyAttribute(BindingSourceKey bindingSource)
```
Why add this? `[FromForm]` does the needful on its own. I don't know why we'd introduce more ways to do the same thing, and it's pretty rare to use a binding source in razor pages.

pranavkm commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/BindingSourceKey.cs:
```diff
@@ -0,0 +1,69 @@
+
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+
+namespace Microsoft.AspNetCore.Mvc
```
Could this be in the `ModelBinding` namespace? I imagine it's not super common for users to need this, no?

dougbu replied:
Leaving as is because `[BindProperty]` and `[ModelBinder]` are in the top-level Mvc namespace, meaning an extra `using` would be needed

pranavkm reviewed (approved):
Looks good to me once @rynowak's happy

dougbu commented on src/Mvc/src/Microsoft.AspNetCore.Mvc.Core/ModelBinderAttribute.cs:
```diff
@@ -26,39 +28,108 @@ namespace Microsoft.AspNetCore.Mvc
+        /// query string, instead use the <see cref="ModelBinderAttribute(BindingSourceKey)"/> constructor to set
+        /// <see cref="BindingSource"/> and set the <see cref="BinderType"/> property.
+        /// </remarks>
         public ModelBinderAttribute(Type binderType)
```
I did _not_ add a `ModelBinderAttribute(Type binderType)` because this constructor already looks redundant to me because `BinderType` is settable. LMK if you want that added.

dougbu committed:
Update `package.json` and `package-lock.json` files

dougbu committed:
Add `BindingSourceKeyExtensions` to centralize `BindingSource` lookup

dougbu commented:
:up::date: to address PR comments (and rebase)

@BrennanConroy and @jkotalik, why are we checking in `package-lock.json` files -- aren't they regenerated on every build?

dougbu commented:
Bump @rynowak

dougbu commented:
@rynowak do you also approve?

rynowak commented:
No, sorry I don't any understanding of why all of this new API and UX is needed. We have too many attributes and things to configure already. I'm not happy to see more alternatives.

dougbu commented:
@rynowak let's chat when you're next in the office

dougbu committed:
Address offline PR comments: Remove parts of this change
- do not change default `BindingSource` used when `BinderType` is set
  - could cause custom binders to not be called
- do not add `BindingSourceKey` or attribute constructor overloads using it
  - provided a second way to do something that was already supported
  - instead document subclassing the attributes when necessary

nits:
- shorten a few added lines

dougbu committed:
Revert "Update `package.json` and `package-lock.json` files"

This reverts commit 3a8b36e8333cee190430af5b0d668b39150af892.

dougbu committed:
Add tests of `BindingSource.ModelBinding`

dougbu commented:
:up::date: to remove many changes and catch up with latest MVC file layout

# What remains
smaller issues:
* catch invalid `BinderType` values up front
* complete `BindingSource.ModelBinding` implementation: `IValueProvider` filtering was faulty

nits:
* accept VS suggestions e.g. remove unused variables
* add `[BindProperty]` doc comments
* add `<remarks>` to `BinderType` properties that recommend setting `BindingSource` in some cases
* "model binder" -> `<see cref="IModelBinder" /> implementation` in some doc comments

dougbu committed:
Add regression test for #4939
